File: python/sliding.py
"""
Greedy sliding window aggregation algorithm.

Reference:
  D. Basin, F. Klaedtke, and E. Zalinescu.
  Greedily Computing Associative Aggregations on Sliding Windows.
  Information Processing Letters, 115(2):186-192, 2015.

Go channels → Python iterators/generators
==========================================
Go's SlidingWindow function receives stream elements from an `in` channel and
pushes results to an `out` channel.  In Python the idiomatic equivalent is a
generator: the caller iterates over the stream with any iterable (list, generator
expression, file, network source, …) and sliding_window itself is a generator
that yields one aggregated value per window, consuming stream elements lazily.

This means:
  - The input stream can be infinite (only elements actually needed are consumed).
  - The result sequence can be consumed lazily by the caller.
  - No threads or async machinery are required.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Generic, Iterable, Iterator, NamedTuple, Optional, TypeVar, Union

T = TypeVar("T")
Op = Callable[[T, T], T]


# ---------------------------------------------------------------------------
# Window
# ---------------------------------------------------------------------------

class Window(NamedTuple):
    left: int
    right: int


# ---------------------------------------------------------------------------
# Lifted operator  (None plays the role of the missing/discharged value)
# ---------------------------------------------------------------------------

def lift(op: Op[T]) -> Op[Optional[T]]:
    """Lift a binary operator so it works on Optional values."""
    def lifted(x: Optional[T], y: Optional[T]) -> Optional[T]:
        if x is None or y is None:
            return None
        return op(x, y)
    return lifted


# ---------------------------------------------------------------------------
# Tree  —  sealed hierarchy: _Leaf | _Node
#
# Splitting into two classes eliminates all Optional fields on the tree
# structure itself:
#   - _Leaf carries no data at all.
#   - _Node always has a label and two (non-optional) children.
# The only remaining Optional is _Node.aggregation, which is legitimately
# optional: None means the node has been discharged (its cached value
# invalidated) while its children are still reusable.
# ---------------------------------------------------------------------------

@dataclass(slots=True)
class _Leaf(Generic[T]):
    """A leaf node carrying no data."""
    def left_index(self) -> int:  return -1
    def right_index(self) -> int: return -1


@dataclass(slots=True)
class _Node(Generic[T]):
    """An inner node with a label and two children."""
    from_idx:    int
    to_idx:      int
    aggregation: Optional[T]   # None when discharged
    left:        Tree[T]
    right:       Tree[T]

    def left_index(self) -> int:  return self.from_idx
    def right_index(self) -> int: return self.to_idx

    def discharge(self) -> None:
        """Invalidate the cached aggregation while keeping children intact."""
        self.aggregation = None


Tree = Union[_Leaf[T], _Node[T]]


# ---------------------------------------------------------------------------
# Constructors
# ---------------------------------------------------------------------------

def _leaf() -> _Leaf:
    return _Leaf()


def _singleton(i: int, x: T) -> _Node[T]:
    lf = _leaf()
    return _Node(from_idx=i, to_idx=i, aggregation=x, left=lf, right=lf)


# ---------------------------------------------------------------------------
# Tree helpers
# ---------------------------------------------------------------------------

def _combine(op: Op[Optional[T]], t1: Tree[T], t2: Tree[T]) -> Tree[T]:
    """Merge two trees under a new node, discharging t1."""
    if isinstance(t1, _Leaf):
        return t2
    if isinstance(t2, _Leaf):
        return t1
    v = op(t1.aggregation, t2.aggregation)
    t1.discharge()
    return _Node(
        from_idx=t1.from_idx,
        to_idx=t2.to_idx,
        aggregation=v,
        left=t1,
        right=t2,
    )


def _reusables(op: Op[Optional[T]], t: Tree[T], i: int, acc: Tree[T]) -> Tree[T]:
    """
    Fold every maximal subtree of `t` whose index range lies entirely at or
    after `i` into `acc` via `_combine`.  Tail-recursive loop (mirrors Go).
    """
    while True:
        if i > t.right_index():
            return acc
        if i == t.left_index():
            return _combine(op, t, acc)
        assert isinstance(t, _Node)
        if i >= t.right.left_index():
            t = t.right   # tail call: _reusables(op, t.right, i, acc)
        else:
            acc = _combine(op, t.right, acc)
            t = t.left    # tail call: _reusables(op, t.left, i, acc)


def _news(op: Op[Optional[T]], stream: Iterator[T], start: int, n: int,
          acc: Tree[T]) -> Tree[T]:
    """
    Read `n` elements from `stream` starting at index `start`, build singleton
    trees, and fold them right-to-left into `acc`.

    The recursion reads the element at `start` first, then recurses for the
    remaining n-1 elements, and combines on the way back up — so the leftmost
    element ends up deepest on the left spine without any intermediate list.

    Equivalent to Go's `news` function; StopIteration propagates to the caller
    (sliding_window) to signal stream exhaustion.

    Note: the recursion depth equals n, the number of new elements in the
    current window step.  Python's default limit is 1000 (adjustable via
    sys.setrecursionlimit).  If window steps can introduce more than ~1000 new
    elements, replace the recursive implementation with the iterative one below,
    which is equivalent but avoids deep call stacks at the cost of an
    intermediate list and a reversal:

        nodes: list[_Node[T]] = []
        for j in range(n):
            nodes.append(_singleton(start + j, next(stream)))
        for node in reversed(nodes):
            acc = _combine(op, node, acc)
        return acc
    """
    if n <= 0:
        return acc
    v = next(stream)  # raises StopIteration when stream is exhausted
    acc = _news(op, stream, start + 1, n - 1, acc)
    return _combine(op, _singleton(start, v), acc)


def _skip(stream: Iterator, n: int) -> None:
    """Discard the next `n` elements from `stream`."""
    for _ in range(n):
        next(stream)  # raises StopIteration on premature exhaustion


def _slide(op: Op[Optional[T]], stream: Iterator[T], t: Tree[T],
           w: Window) -> Tree[T]:
    """Advance tree `t` by one window step, consuming elements from `stream`."""
    from_idx = max(w.left, 1 + t.right_index())
    to_idx = w.right
    # Skip elements that fall between the previous window and the current one.
    _skip(stream, from_idx - (1 + t.right_index()))
    # Fold newly seen elements.
    r = _news(op, stream, from_idx, max(0, to_idx - from_idx + 1), _leaf())
    # Fold reusable subtrees from the previous window.
    return _reusables(op, t, w.left, r)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def sliding_window(
    op: Op[T],
    stream: Iterable[T],
    windows: Iterable[Window],
) -> Iterator[T]:
    """
    Compute the aggregation of stream elements within a sliding window.

    Parameters
    ----------
    op      : associative binary operator  (a, b) -> a op b
    stream  : iterable of data elements x_0, x_1, x_2, ...
              (may be an infinite generator; only elements actually required
              by the windows are consumed)
    windows : iterable of Window(left, right) pairs satisfying:
                0 <= l_0 <= l_1 <= ...
                0 <= r_0 <= r_1 <= ...
                l_i <= r_i  for all i

    Yields
    ------
    y_i = x[l_i] op x[l_i+1] op ... op x[r_i]  for each window i

    The function is itself a generator, so results are produced lazily.
    """
    lop: Op[Optional[T]] = lift(op)
    it: Iterator[T] = iter(stream)
    t: Tree[T] = _leaf()
    try:
        for w in windows:
            t = _slide(lop, it, t, w)
            assert isinstance(t, _Node) and t.aggregation is not None, \
                "no aggregated value at tree's root"
            yield t.aggregation
    except StopIteration:
        # Stream exhausted before all windows were satisfied; stop silently,
        # matching Go's behaviour of returning without sending further results.
        return

File: typescript/sliding.ts
/**
 * Sliding Window Algorithm — TypeScript Implementation
 *
 * Stream abstraction
 * ------------------
 * The Go implementation uses channels (chan A) for both input and output.
 * In TypeScript the idiomatic equivalent is AsyncIterable<T>:
 *   - Input  : AsyncIterable<A>  – consumed lazily, element by element.
 *   - Output : AsyncGenerator<A> – produced lazily via `yield`.
 * This supports infinite streams and composes naturally with
 * `for await...of`.  If all data is already in memory a plain Array
 * (which is a synchronous Iterable) can be wrapped with a trivial
 * async generator, see the `fromArray` helper at the bottom.
 */

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

/** A binary associative operator over values of type A. */
export type Op<A> = (x: A, y: A) => A;

/** A sliding window over a data stream.  Both bounds are 0-based indices. */
export interface Window {
  left: number;
  right: number;
}

// ---------------------------------------------------------------------------
// Internal: option type
// ---------------------------------------------------------------------------

type Option<A> = { ok: true; value: A } | { ok: false };

function some<A>(v: A): Option<A> { return { ok: true, value: v }; }
function none<A>(): Option<A>     { return { ok: false }; }

function lift<A>(op: Op<A>): Op<Option<A>> {
  return (x, y) =>
    x.ok && y.ok ? some(op(x.value, y.value)) : none<A>();
}

// ---------------------------------------------------------------------------
// Internal: tree
// ---------------------------------------------------------------------------

interface Label<A> {
  from: number;
  to: number;
  agg: Option<A>;   // aggregated value [from..to]; none when discharged
}

type Tree<A> = null | { data: Label<A>; left: Tree<A>; right: Tree<A> };

// Constructors

function leafTree<A>(): Tree<A> { return null; }

function singleton<A>(i: number, x: A): Tree<A> {
  return { data: { from: i, to: i, agg: some(x) }, left: null, right: null };
}

/** Combine two trees under a new root.  `t1` is discharged (its aggregation
 *  cleared) and becomes the left child; `t2` becomes the right child.
 *  Returns the non-null tree unchanged when one side is a leaf. */
function combine<A>(op: Op<Option<A>>, t1: Tree<A>, t2: Tree<A>): Tree<A> {
  if (t1 === null) return t2;
  if (t2 === null) return t1;
  return {
    data: {
      from: t1.data.from,
      to:   t2.data.to,
      agg:  op(t1.data.agg, t2.data.agg),
    },
    left:  discharge(t1),
    right: t2,
  };
}

/** Clear the aggregation of the root node (non-destructive on the rest). */
function discharge<A>(t: NonNullable<Tree<A>>): Tree<A> {
  return { data: { from: t.data.from, to: t.data.to, agg: none<A>() },
           left: t.left, right: t.right };
}

// Selectors

function rightIndex<A>(t: Tree<A>): number {
  return t === null ? -1 : t.data.to;
}

function leftIndex<A>(t: NonNullable<Tree<A>>): number {
  return t.data.from;
}

function extract<A>(t: Tree<A>): A {
  if (t === null || !t.data.agg.ok) {
    throw new Error("no aggregated value at tree root");
  }
  return t.data.agg.value;
}

// ---------------------------------------------------------------------------
// Internal: tree-building helpers
// ---------------------------------------------------------------------------

/** Read `n` elements from the async-iterator starting at index `i`, build
 *  singleton trees right-to-left, and combine them into `acc`.
 *  Returns the updated accumulator tree, or null if the stream ended early. */
async function readNew<A>(
  op: Op<Option<A>>,
  iter: AsyncIterator<A>,
  i: number,
  n: number,
  acc: Tree<A>,
): Promise<Tree<A> | null> {
  if (n <= 0) return acc;

  const { value, done } = await iter.next();
  if (done) return null;                    // stream ended early

  // Recurse first so elements are accumulated right-to-left (deepest-left
  // spine), matching the Haskell/Go behaviour.
  const acc2 = await readNew(op, iter, i + 1, n - 1, acc);
  if (acc2 === null) return null;

  return combine(op, singleton(i, value), acc2);
}

/** Fold every maximal subtree of `t` whose index range lies entirely at or
 *  after `l` into `acc` via `combine`.  Iterative to avoid stack overflow
 *  on deep trees (mirrors the Go implementation). */
function reusables<A>(
  op: Op<Option<A>>,
  t: Tree<A>,
  l: number,
  acc: Tree<A>,
): Tree<A> {
  while (true) {
    if (t === null || l > t.data.to) return acc;
    if (l === t.data.from)           return combine(op, t, acc);

    const { left: t1, right: t2 } = t;
    if (t2 !== null && l >= t2.data.from) {
      t = t2;                          // tail call: reusables(op, t2, l, acc)
    } else {
      acc = combine(op, t2, acc);
      t   = t1;                        // tail call: reusables(op, t1, l, acc)
    }
  }
}

/** Skip `k` elements from the async iterator. Returns false if the stream
 *  ended before `k` elements were consumed. */
async function skip<A>(iter: AsyncIterator<A>, k: number): Promise<boolean> {
  for (let i = 0; i < k; i++) {
    const { done } = await iter.next();
    if (done) return false;
  }
  return true;
}

// ---------------------------------------------------------------------------
// Internal: slide
// ---------------------------------------------------------------------------

/** Advance the aggregation tree by one window step.
 *  Returns the updated tree or null if the input stream ended early. */
async function slide<A>(
  op: Op<Option<A>>,
  iter: AsyncIterator<A>,
  t: Tree<A>,
  w: Window,
): Promise<Tree<A> | null> {
  const from = Math.max(w.left, 1 + rightIndex(t));
  const to   = w.right;

  // Discard elements that fall between the previous and the current window.
  const skipped = from - (1 + rightIndex(t));
  if (skipped > 0 && !(await skip(iter, skipped))) return null;

  // Fold newly received elements (those not in the previous window).
  const newCount = Math.max(0, to - from + 1);
  const r = await readNew(op, iter, from, newCount, leafTree<A>());
  if (r === null) return null;

  // Fold reusable subtrees from the previous window.
  return reusables(op, t, w.left, r);
}

// ---------------------------------------------------------------------------
// Public: slidingWindow
// ---------------------------------------------------------------------------

/**
 * Compute the aggregations of stream elements within a sliding window for
 * an associative operator.
 *
 * @param input   An `AsyncIterable` (or plain `Iterable`) delivering the
 *                data stream x₀, x₁, x₂, … in order.  Elements are consumed
 *                lazily: only as many as required by the windows are read.
 * @param windows An `Iterable` (or `AsyncIterable`) of windows.  Windows must
 *                satisfy  0 ≤ l₀ ≤ l₁ ≤ …,  0 ≤ r₀ ≤ r₁ ≤ …,  lᵢ ≤ rᵢ.
 * @param op      A binary associative operator.
 *
 * @returns An `AsyncGenerator` that yields yᵢ = x[lᵢ] op x[lᵢ+1] op … op x[rᵢ]
 *          for each window.  Terminates early if the input stream closes.
 *
 * @example
 * // Sum over a fixed-size sliding window of width 3
 * const data    = fromArray([1, 2, 3, 4, 5, 6]);
 * const windows = fromArray([
 *   { left: 0, right: 2 },
 *   { left: 1, right: 3 },
 *   { left: 2, right: 4 },
 *   { left: 3, right: 5 },
 * ]);
 * for await (const result of slidingWindow(data, windows, (a, b) => a + b)) {
 *   console.log(result); // 6, 9, 12, 15
 * }
 */
export async function* slidingWindow<A>(
  input:   AsyncIterable<A> | Iterable<A>,
  windows: AsyncIterable<Window> | Iterable<Window>,
  op: Op<A>,
): AsyncGenerator<A> {
  const lop  = lift(op);
  const iter = toAsyncIterator(input);
  let   t: Tree<A> = leafTree<A>();

  for await (const w of windows) {
    const t2 = await slide(lop, iter, t, w);
    if (t2 === null) return;   // input stream exhausted
    t = t2;
    yield extract(t);
  }
}

// ---------------------------------------------------------------------------
// Helpers for callers
// ---------------------------------------------------------------------------

/** Wrap a plain array (or any synchronous iterable) as an AsyncIterable. */
export async function* fromArray<A>(xs: Iterable<A>): AsyncIterable<A> {
  yield* xs;
}

/** Obtain an AsyncIterator from either a sync or async iterable. */
function toAsyncIterator<A>(
  source: AsyncIterable<A> | Iterable<A>,
): AsyncIterator<A> {
  if (Symbol.asyncIterator in source) {
    return (source as AsyncIterable<A>)[Symbol.asyncIterator]();
  }
  return (async function* () { yield* source as Iterable<A>; })()[Symbol.asyncIterator]();
}
